state/watcher/hubwatcher.go

```diff
@@ -0,0 +1,327 @@
+	w.tomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *HubWatcher) Wait() error {
+	return w.tomb.Wait()
+}
+
+// Stop stops all the watcher activities.
+func (w *HubWatcher) Stop() error {
+	return worker.Stop(w)
+}
+
+// Dead returns a channel that is closed when the watcher has stopped.
+func (w *HubWatcher) Dead() <-chan struct{} {
+	return w.tomb.Dead()
+}
+
+// Err returns the error with which the watcher stopped.
+// It returns nil if the watcher stopped cleanly, tomb.ErrStillAlive
+// if the watcher is still running properly, or the respective error
+// if the watcher is terminating or has terminated with an error.
+func (w *HubWatcher) Err() error {
+	return w.tomb.Err()
+}
+
+func (w *HubWatcher) sendReq(req interface{}) {
+	select {
+	case w.request <- req:
+	case <-w.tomb.Dying():
+	}
+}
+
+// Watch starts watching the given collection and document id.
+// An event will be sent onto ch whenever a matching document's txn-revno
+// field is observed to change after a transaction is applied. The revno
+// parameter holds the currently known revision number for the document.
+// Non-existent documents are represented by a -1 revno.
+func (w *HubWatcher) Watch(collection string, id interface{}, revno int64, ch chan<- Change) {
+	if id == nil {
+		panic("watcher: cannot watch a document with nil id")
+	}
+	w.sendReq(reqWatch{watchKey{collection, id}, watchInfo{ch, revno, nil}})
+}
+
+// WatchCollection starts watching the given collection.
+// An event will be sent onto ch whenever the txn-revno field is observed
+// to change after a transaction is applied for any document in the collection.
+func (w *HubWatcher) WatchCollection(collection string, ch chan<- Change) {
+	w.WatchCollectionWithFilter(collection, ch, nil)
+}
+
+// WatchCollectionWithFilter starts watching the given collection.
+// An event will be sent onto ch whenever the txn-revno field is observed
+// to change after a transaction is applied for any document in the collection, so long as the
+// specified filter function returns true when called with the document id value.
+func (w *HubWatcher) WatchCollectionWithFilter(collection string, ch chan<- Change, filter func(interface{}) bool) {
+	w.sendReq(reqWatch{watchKey{collection, nil}, watchInfo{ch, 0, filter}})
+}
+
+// Unwatch stops watching the given collection and document id via ch.
+func (w *HubWatcher) Unwatch(collection string, id interface{}, ch chan<- Change) {
+	if id == nil {
+		panic("watcher: cannot unwatch a document with nil id")
+	}
+	w.sendReq(reqUnwatch{watchKey{collection, id}, ch})
+}
+
+// UnwatchCollection stops watching the given collection via ch.
+func (w *HubWatcher) UnwatchCollection(collection string, ch chan<- Change) {
+	w.sendReq(reqUnwatch{watchKey{collection, nil}, ch})
+}
+
+// loop implements the main watcher loop.
+// period is the delay between each sync.
+func (w *HubWatcher) loop() error {
+	for {
+		w.logger.Warningf("loop() select")
```

Comment:
Warningf?

Reply:
Yeah... I would have marked them critical but my logger interface defined didn't have it. These were temporary logging that should be removed.